torch/_subclasses/fake_tensor.py

```diff
@@ -1414,6 +1524,215 @@ def __exit__(self, a, b, c):
+            # The disallowance of dynamic shapes can affect caching, e.g., a change of
+            # mode can introduce a DynamicOutputShapeException where it wasn't seen on
+            # a previous instance of the same func and arg combination.
+            self.shape_env and self.shape_env.allow_dynamic_output_shape_ops,
+        )
+        return _DispatchCacheKey(key_values)
+
+    def _translate_arg(self, arg: Any) -> Any:
+        """
+        Translate the provided arg into a form suitable for caching at FakeTensor
+        dispatch, i.e., convert unhashable types like lists & dicts into tuples and
+        convert FakeTensors into metadata. Raises _BypassDispatchCache to signal
+        unsupported cases that should bypass caching.
+        """
+        if isinstance(arg, (list, tuple)):
+            return tuple(self._translate_arg(x) for x in arg)
+        elif isinstance(arg, dict):
+            return tuple((k, self._translate_arg(v)) for k, v in arg.items())
+        elif isinstance(arg, (torch.SymBool, torch.SymInt, torch.SymFloat)):
+            raise _BypassDispatchCache("symbolic shape")
+        elif isinstance(arg, FakeTensor):
+            if not self.is_our_fake(arg):
+                raise _BypassDispatchCache("not our fake")
+            if arg._has_symbolic_sizes_strides:
+                raise _BypassDispatchCache("symbolic shape")
+            if arg.constant is not None:
+                raise _BypassDispatchCache("constant attribute")
+            return extract_tensor_metadata(arg)
+        elif isinstance(arg, torch.Tensor):
+            raise _BypassDispatchCache("non-fake tensor")
+        else:
+            # It's important to capture the type of the arg since, e.g., 1 and 1.0
+            # hash to the same value, but can produce different dtypes for the
+            # output tensor.
+            return (type(arg), arg)
+
+    def _create_cache_entry(
+        self, key: _DispatchCacheKey, args: Tuple[Any], output: FakeTensor
+    ) -> _DispatchCacheEntry:
+        """
+        Create an entry in the cache for the given 'output' Tensor.
+        """
+        # Some ops return tuples of Tensors, but it's rare, so avoid
+        # the complexity of caching other types.
+        if not isinstance(output, FakeTensor):
+            raise _BypassDispatchCache("non-FakeTensor output")
+
+        # Avoid caching FakeTensors with constants attached since those
+        # can be invalidated.
+        if output.constant is not None:
+            raise _BypassDispatchCache("constant attribute")
+
+        # TODO: support caching sparse outputs?
+        if output.is_sparse:
+            raise _BypassDispatchCache("sparse output")
+
+        # If this is an in-place op, we just record which input arg is aliased.
+        entry = None
+        for idx in range(len(args)):
+            if id(args[idx]) == id(output):
+                entry = _DispatchCacheEntry(None, idx)
+                break
+
```

Comment:
I think we need to handle `kwargs` as well. And I wonder if there's any nested data structures that matter here..

Reply:
@eellison Oh really? Do you know of any ops that alias one of the kwarg inputs that I can play with?